Improve Prometheus Metrics Support in Helm Chart

EladAviczer opened the pull request:
Adding a new ServiceMonitor template under templates/metrics/ to expose controller manager metrics to Prometheus.

Introducing configurable options in values.yaml to enable/disable the ServiceMonitor, set its namespace, and apply custom labels.

Updating values.schema.json to validate the new configuration options.

Removing the old ServiceMonitor template from templates/prometheus/ for better structure and clarity.
#544

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/k6-operator?pullRequest=548) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/k6-operator?pullRequest=548) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/k6-operator?pullRequest=548) it.</sub>

EladAviczer commented:
@yorugac : )

yorugac reviewed (changes requested):
Hi @EladAviczer, thank you for the PR!

It's working :raised_hands: But I have a couple of requests to improve changes a bit. With this change, `prometheus` section is still present in `values.yaml` even though it's meaningless now. So it should be removed in order not to cause false positive for someone. 
Btw, I'll need to mark this PR as a breaking change. Stands to reason since this is about renaming a whole section.

Could you please also add a small sample for new ServiceMonitor configuration? We don't have much Helm samples now but we should try to improve that and this seems like a good case for that :smile: So an additional file in `charts/k6-operator/samples` with some random values:
```yaml
metrics:
  serviceMonitor:
    enabled: true
    namespace: monitoring
    jobLabel: prometheus-operator
    selector:
     ...
```
This sample is for illustrative purposes only (at least for now).

I have also left a couple of comments in the code: please see below.

EladAviczer commented on charts/k6-operator/values.schema.json:
```diff
@@ -1,4 +1,6 @@
 {
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
```
Actually got LLM help here because i couldn't helm template to debug without this change...
i guess i will restore the file to the origin state okay ?

EladAviczer replied:
got help from LLM here :)
will use your makefile instead

EladAviczer replied:
Okay used your makefile and everything works nice now, found some bugs in the values file thanks to it and fixed those too.

btw i will add this schema to my charts.  thanks :)

EladAviczer replied:
fixed it now, please recheck :)

EladAviczer commented:
Fixed everything.
About the 

> `prometheus` section is still present in `values.yaml`

 from what i understand i already deleted it before PRing.
Please recheck :)

yorugac commented on charts/k6-operator/values.schema.json:
```diff
@@ -311,19 +386,6 @@
       "title": "podLabels",
       "type": "object"
     },
-    "prometheus": {
```
@EladAviczer, I guess the confusion came from the `prometheus` section in here, in schema. As in, yesterday I could create values referencing `prometheus` and it passed validation. Now section is removed from schema and it doesn't pass it :slightly_smiling_face:

yorugac commented on charts/k6-operator/values.yaml:
```diff
@@ -116,13 +116,84 @@ namespace:
-  enabled: false
+  serviceMonitor:
+    # @schema
+    # required: false
```
```suggestion
    # required: true
```
Maybe this is a nit, but if one adds `metrics.serviceMonitor`, they probably should also set `enabled`, otherwise there won't be any point to it.

yorugac reviewed (changes requested):
Hi @EladAviczer, thanks for the update! Looks great and works, including validation now. (I agree that the schema is pretty neat! Esp. with all the help from community to fix & improve it :joy: )
Just one nit pick left: I suggest to change `metrics.serviceMonitor.enabled` to required in validation, because if the person is defining ServiceMonitor but not setting `enabled`, then probably they made a mistake. So ~ validation as early catching of typos. WDYT?

EladAviczer commented:
> Hi @EladAviczer, thanks for the update! Looks great and works, including validation now. (I agree that the schema is pretty neat! Esp. with all the help from community to fix & improve it 😂 ) Just one nit pick left: I suggest to change `metrics.serviceMonitor.enabled` to required in validation, because if the person is defining ServiceMonitor but not setting `enabled`, then probably they made a mistake. So ~ validation as early catching of typos. WDYT?

Nice!
commited your suggestions, it seem neat!

yorugac reviewed (approved):
LGTM; thanks for working on this, @EladAviczer :raised_hands: 

For the record. Despite the addition of `required` to `metrics.serviceMonitor.enabled` field, it _is_ possible to create a chart with the following:
```yaml
metrics:
  serviceMonitor:
    namespace: monitoring
```
It will not contain ServiceMonitor because `enabled` isn't set, yet validation doesn't complain for some reason :confused: Not sure why or how to fix it ATM. (other than adding all possible fields into IF statement which is obviously an overkill) This needs digging. But here's the warning :point_up: to whoever stumbles on this.
